pytest/tests/loadtest/locust/common/base.py

```diff
@@ -364,6 +370,98 @@ def prepare_account(self, account: Account, parent: Account, balance: int,
         account.refresh_nonce(self.node)
         return exists
 
+    def prepare_accounts(self,
+                         accounts: typing.List[Account],
+                         parent: Account,
+                         balance: int,
+                         msg: str,
+                         timeout: timedelta = timedelta(minutes=3)):
+        """
+        Creates accounts if they don't exist and refreshes their nonce.
+        Accounts must share the parent account.
+        
+        This implementation attempts on-chain parallelization, hence it should
+        be faster than calling `prepare_account` in a loop.
+        
+        Note that error-handling in this variant isn't quite as smooth. Errors
+        that are only reported by the sync API of RPC nodes will not be caught
+        here. Instead, we do a best-effort retry and stop after a fixed timeout.
+        """
+
+        # To avoid blocking, each account goes though a FSM independently.
+        #
+        # FSM outline:
+        #
+        #    [INIT] -----------------(account exists already)-------------
+        #       |                                                        |
+        #       V                                                        V
+        #  [TO_CREATE] --(post tx)--> [INFLIGHT] --(poll result)--> [TO_REFRESH]
+        #   ^                    |                                       |
+        #   |                    |                                       |
+        #   |--(fail to submit)<--                                   (refresh)
+        #                                                                |
+        #                                                                V
+        #                                                              [DONE]
+        #
```

Comment:
I commend your non-lazy approach to writing a whole async state machine! :)

If you do want to lean on the lazy side, submitting multiple calls to `prepare_account` into [ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor) might be a viable alternative that should still give you a speed up with less code. Taking this a bit further, https://docs.python.org/3/library/asyncio.html would be a Python-recommended way to implement such state machines (given that you don't batch any submitted transactions or query them in a batch) - that might be a bit more verbose but will allow more than CPU-count number of in-flight requests.

Reply:
Thanks for the ideas and links, really appreciated!

Hm, I was hoping to not introduce another layer of concurrency frameworks. We already have locust users (greenlets), as well as multiple worker instances executing concurrently.

My approach is a bit more code and maybe not trivial code, but I would argue from an architectural perspective it is simpler  than a thread pool because it's all sequential code. And because the state machine here seems like relevant business logic, I'm fine with coding it out by hand rather than relying on some "automatic" state machine.

To make my argument a bit less abstract: Each account creation transaction reserves a Nonce from the parent account, a global-per-worker variable. The transactions must be sent to the chain in the same order as those nonces, or else we will have race conditions where they invalidate each other if they aren't all included in the same chunk. That would lead to failures that need to be handled by updating the nonce, a failure condition I don't think is possible to occur in my code.

But let's say I do see some invalid transactions from this code. Right now, I can easily debug it and I know in what order transactions are sent within a single user because it's all sequential code. It's still hairy when I think about it for multiple users/greenlets, but for now this is only used in setup code so that's not a real problem. However, this will change soon for the sweat stress test where we currently create 1000 accounts in a naive & sequential loop, I want to replace that with this code, too.

Either way, multiply layers of concurrency multiplies the system complexity, whereas a well-documented manual state machine keeps complexity local and presents a simple and composable interface.

All that said, I might still fall back to this in order to make it more generally applicable for things like FT contract installation. I'll have to play around with it a bit...